chips/sam4l/src/pm.rs

```diff
@@ -259,243 +148,170 @@ pub enum SystemClockSource {
     /// Use the internal digital frequency locked loop (DFLL) sourced from
     /// the internal RC32K clock. Note this typically requires calibration
     /// of the RC32K to have a consistent clock.
-    DfllRc32k,
+    DfllRc32kAt48MHz,
 
     /// Use an external crystal oscillator as the direct source for the
     /// system clock.
-    ExternalOscillator,
+    ExternalOscillatorAt16MHz,
 
     /// Use an external crystal oscillator as the input to the internal
     /// PLL for the system clock. This expects a 16 MHz crystal.
-    ExternalOscillatorPll,
+    PllExternalOscillatorAt48MHz,
+}
+
+/// Which frequency range does your external oscillator fall in? Configuration
+/// needs to change based on this.
+pub enum OscClock {
+    /// 16 MHz external oscillator
+    Frequency16MHz,
+}
+
+/// Configuration for the startup time of the external oscillator. In practice
+/// we have found that some boards work with a short startup time, while others
+/// need a slow start in order to properly wake from sleep.
+#[derive(Copy,Clone,Debug)]
+pub enum OscStartupMode {
+    /// Use a fast startup. ~0.5 ms in practice.
+    FastStart,
+
+    /// Use a slow startup. ~8.9 ms in practice.
+    SlowStart,
 }
 
 const PM_BASE: usize = 0x400E0000;
-const BSCIF_BASE: usize = 0x400F0400;
-const SCIF_BASE: usize = 0x400E0800;
-const FLASHCALW_BASE: usize = 0x400A0000;
 
 const HSB_MASK_OFFSET: u32 = 0x24;
 const PBA_MASK_OFFSET: u32 = 0x28;
 const PBB_MASK_OFFSET: u32 = 0x2C;
 const PBC_MASK_OFFSET: u32 = 0x30;
 const PBD_MASK_OFFSET: u32 = 0x34;
 
-static mut PM: *mut PmRegisters = PM_BASE as *mut PmRegisters;
-static mut BSCIF: *mut BscifRegisters = BSCIF_BASE as *mut BscifRegisters;
-static mut SCIF: *mut ScifRegisters = SCIF_BASE as *mut ScifRegisters;
-static mut FLASHCALW: *mut FlashcalwRegisters = FLASHCALW_BASE as *mut FlashcalwRegisters;
-
-static mut SYSTEM_FREQUENCY: VolatileCell<u32> = VolatileCell::new(0);
+static mut PM_REGS: *mut PmRegisters = PM_BASE as *mut PmRegisters;
+pub static mut PM: PowerManager = PowerManager::new();
 
-unsafe fn unlock(register_offset: u32) {
-    (*PM).unlock.set(0xAA000000 | register_offset);
+pub struct PowerManager {
+    system_frequency: Cell<u32>,
+    oscillator_frequency: Cell<u32>,
+    oscillator_startup: Cell<OscStartupMode>,
 }
 
-unsafe fn select_main_clock(clock: MainClock) {
-    unlock(0);
-    (*PM).mcctrl.set(clock as u32);
-}
+impl PowerManager {
+    const fn new() -> PowerManager {
+        // initialize with default values
+        PowerManager {
+            system_frequency: Cell::new(0),
+            oscillator_frequency: Cell::new(0),
+            oscillator_startup: Cell::new(OscStartupMode::FastStart),
+        }
+    }
 
-/// Enable HCACHE
-unsafe fn enable_cache() {
-    enable_clock(Clock::HSB(HSBClock::FLASHCALWP));
-    enable_clock(Clock::PBB(PBBClock::HRAMC1));
-    // Enable cache
-    (*FLASHCALW).ctrl.set(0x01);
+    pub unsafe fn specify_external_oscillator(&self,
```

Comment:
I'm confused what you mean. This method is for setting the `oscillator_startup` field so that we can later modify our system clock configuration based on it. The reason the field exists in the first place is so that we can store the setting from here for later use.